### PR TITLE
SDSTOR-9150 add new homestore data service apis

### PR DESCRIPTION
**Pull request by yamingk**
1. Add a new API for data service that will take block ids as input for async_write;
2. add doxygen comment for public APIs
3. new api was tested by home replication storage test binary;

### CODE REVIEW

File: src/include/homestore/blkdata_service.hpp [hunk 1]
```diff
@@ -52,59 +52,120 @@ class BlkDataService {
...
+
+    /**
+     * @brief : asynchronous read
+     *
+     * @param bid : block id to read
+     * @param sgs : the read buffer stored
+     * @param size : size to read
+     * @param cb : callback that will be triggered after read completes
+     * @param part_of_batch : is this read part of batch;
+     */
```
Comment (hkadayam) [inline src/include/homestore/blkdata_service.hpp:99]: My suggestion is to rename `async_write_ahead()` to `async_write()` and current `async_write()` to `async_alloc_write()`. Write ahead in general has different meaning right (similar to read_ahead)?

Comment (yamingk) [reply src/include/homestore/blkdata_service.hpp:99]: I will take your suggestion, let me update the PR.

**Review by hkadayam — APPROVED** [review]
Looks good

File: src/include/homestore/homestore.hpp [hunk 2]
```diff
@@ -132,14 +132,12 @@ class HomeStore {
...
-    HomeStore& with_data_service(float size_pct);
-    HomeStore& with_log_service(float data_size_pct, float ctrl_size_pct);
-    HomeStore& with_index_service(float index_size_pct, std::unique_ptr< IndexServiceCallbacks > cbs);
     HomeStore& after_init_done(hs_init_done_cb_t init_done_cb);
     HomeStore& before_init_devices(hs_init_starting_cb_t init_start_cb);
 
-    void init(bool wait_for_init = false);
+    void init_internal(bool wait_for_init);
+    void init(bool wait_for_init = false, float meta_size_pct = 0, float log_data_size_pct = 0,
+              float log_ctrl_size_pct = 0, float data_svc_size_pct = 0);
```
Comment (hkadayam) [inline src/include/homestore/homestore.hpp:140]: I didn't notice this change, can we please revert this. The idea of `.with` is to be explicit. If you don't agree with it, we can discuss and finalize it?

**Review by hkadayam — COMMENTED** [review]
Can we discuss and finalize the .init() API part. I liked the way explicitly setting up the service. In future we can add more service seamlessly.

**Review by hkadayam — APPROVED** [review]
Looks good to me

**Pull request merged**